calyx/src/passes/well_formed.rs

```diff
@@ -20,18 +23,51 @@ pub struct WellFormed {
     used_groups: HashSet<ir::Id>,
     /// Names of combinational groups used in the control.
     used_comb_groups: HashSet<ir::Id>,
+    /// external cell types of components used in the control.
+    external_cell_types: HashMap<ir::Id, LinkedHashMap<ir::Id, CellType>>,
 }
 
-impl Default for WellFormed {
-    fn default() -> Self {
+impl ConstructVisitor for WellFormed {
+    fn from(ctx: &ir::Context) -> CalyxResult<Self>
+    where
+        Self: Sized,
+    {
         let reserved_names =
             RESERVED_NAMES.iter().map(|s| s.to_string()).collect();
 
-        WellFormed {
+        let mut e_c_t = HashMap::new();
+        for comp in ctx.components.iter() {
+            let mut cellmap: LinkedHashMap<ir::Id, CellType> =
+                LinkedHashMap::new();
+            for cell in comp.cells.iter() {
+                if cell.borrow().is_external() {
+                    if comp.name.eq(&"main") {
```

Comment:
It might be cleaner to do this in two loop nests. One that only checks if `ctx.entrypoint` doesn't have external cells and one that builds the signature for other components.